README.md
# What is Pylene ?


[![pipeline status](https://gitlab.lrde.epita.fr/olena/pylene/badges/master/pipeline.svg)](https://gitlab.lrde.epita.fr/olena/pylene/-/commits/master)
[![coverage report](https://gitlab.lrde.epita.fr/olena/pylene/badges/master/coverage.svg)](https://gitlab.lrde.epita.fr/olena/pylene/-/commits/master)


Pylene is a fork of [Milena](http://www.lrde.epita.fr/olena), an image processing
library targeting genericity and efficiency. It provided mostly Mathematical Morphology building blocs for image
processing pipelines. The library has the following objectives in mind:

* Simplicity: both python bindings and simple C++ syntax
* Efficiency: write algorithms in a simple way and run them as if they were written in C. We follow one guideline: zero-cost abstraction.
* Genericity: write algorithms that are able to run on many kind of images with, yet, zero-cost abstraction.
* Interopability: run pylene algorithms on image coming from external libraries (even on your own image type).


# License

Under [Mozilla Public License, v. 2.0](http://mozilla.org/MPL/2.0/).

# Prerequisited
Pylene is developed in modern C++ so you need a recent C++ compiler. The followings compilers are currently supported
and tested:

* GCC 10, 11, 12
* Clang 12, 13, 14, 15
* ~~Microsoft Visual Studio 2019~~ (Dependencies issue from Conan Center Index)

This project relies on some third part libraries, which are handled by the [Conan Package Manager](https://conan.io/) (version 2.0.1 minimum)

# Quick Start

## Installation

The preferred and strongly recommended way to use Pylene is using Conan, a C++ package manager, and CMake. For using the
library by other means, refer to the [documentation](http://olena.pages.lre.epita.fr/pylene/next/tutorial/installation)

Documentation for all versions can be found [here](http://olena.pages.lre.epita.fr/pylene/).

1. Add to the Conan remote the artifactory of the LRE, where the Pylene artifact is stored:

```
conan remote add lrde-public https://artifactory.lrde.epita.fr/artifactory/api/conan/lrde-public
```

2. At the root of your project, create a Conan recipe (``conanfile.txt``) containing at least the following:

```
[generators]
CMakeDeps
CMakeToolchain

[requires]
pylene/head@lrde/stable
```

3. Create a ``CMakeLists.txt`` containing at least the following to link Pylene to your executable:

```
target_link_libraries(your_target PUBLIC Pylene::core) # Minimal
```

or to execute the following code sample

```
target_link_libraries(your_target PUBLIC Pylene::io-freeimage)
```

4. Configure the project and build it:

```
$ mkdir build && cd build
$ conan install .. -s build_type=Release -s compiler.cppstd=20 -s compiler.libcxx=libstdc++11 --build missing --build openjpeg*
$ cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=./conan_toolchain.cmake
$ cmake --build .
```

For more information or details about the installation process, please refer to the documentation.

## Code sample


To use the algorithms of the library, just include the corresponding headers. Here is an example to perform a grayscale dilation
with a disc of radius 5.

```cpp
#include <mln/core/image/ndimage.hpp>
#include <mln/morpho/dilation.hpp>
#include <mln/core/se/disc.hpp>

#include <mln/io/imread.hpp>
#include <mln/io/imsave.hpp>


int main(int argc, char** argv)
{
  mln::image2d<uint8_t> input;

  mln::io::imread(argv[1], input);
  auto output = mln::morpho::dilation(input, mln::se::disc(5));
  mln::io::imsave(output, argv[2]);
}
```

## What is in there ?

* Image in any dimension (2D, 3D for medical imaging...)
* Image with any value type (RGB-8, 16-bits, hyperspectral images...)

Morphological algorithms:

* se-based filter (dilation, erosion, openings, top-hats, rank-filters...)
* connected filters (area connected openings/closings, attributes filters; grain filters)
* morphological representations (min/max-tree, Tree of Shapes)
* segmentation algorithms (watersheds)


If you want to see more of what is in the library, have a look on the [C++ Documentation](http://olena.pages.lrde.epita.fr/pylene/)


## Why is Pylene singular when compared to other libraries ?

* It is generic: it allows to write a single algorithm that could be apply on all compatibles image types
* It enables lazy-computing with a view system similar to the Boost Generic Image Library

Have a look on the [introduction](http://olena.pages.lrde.epita.fr/pylene/intro)


# Other ressources

Other resources (mind the dates, the library probably has changed since then):

* [A Modern C++ Point of View of Programming in Image Processing (GPCE'22)](https://www.lrde.epita.fr/dload/papers/roynard.22.gpce.pdf)
* [A Modern C++ Library for Generic and Efficient Image Processing (GTGDMM'18)](https://www.lrde.epita.fr/dload/presentations/2018-06-19-geraud.2018.gtgdmmm.pdf)


CHANGELOG.md
# CHANGELOG

## Version 0.2

### Maintainer

* Edwin Carlinet <edwin1.carlinet@epita.fr>
* Baptiste Esteban <baptiste.esteban@epita.fr>
* Michaël Roynard <michael.roynard@epita.fr>

### Contributors

* Philippe Bernet <philippe.bernet@epita.fr>
* Alexandre Bourquelot <alexandre.bourquelot@epita.fr>
* Quentin Kaci <quentin.kaci@epita.fr>
* Victor Simonin <vsimonin@lrde.epita.fr>
* Célian Gossec <cgossec@lrde.epita.fr>

### Features

* Parallel implementation for pointwise algorithms
* SIMD erosion/dilation/transposition
* Partitionning hierarchy :
  * Alphatree
  * Hierachical watersheds
  * Silency maps
  * Horizontal cut
* FITS Image reader
* Multivariate Tree of Shapes
* Scribo module
  * line detection on document images
* Watershed and waterfall based on user input markers
* Hough line transform
* Implement the subtractive rule for component-tree filtering
* Conan v2 compatibility

### Various fixes

* Adding padding facilities into the library
* New transform function signatures taking advantage of concepts
* Fix bug  #72
* Pylene <-> NumPy image conversion
* Update Python module documentation
* Tile type for buffer primitive
* Improve alphatree in the library
* New conan/cmake targets
* Make some fixes in the hierarchical representations of Pylene
* Add compilers in the CI + split Pylene-core into components
* Fix canonization of the alphatree and watershed hierarchy
* Fix the input type checking of the numpy array type caster
* Fix missing traits indexer for signed MM operators.
* Fix Conan deployment after the addition of the Scribo component
* Fix issue #90
* Write documentation about the IO operations
* Remove dependencies on boost::concepts
* Check minimum compiler version for Clang and GCC
* Fix bug in cvxhull impl
* Weighted hough transform
* Update Windows CI and package FreeImage
* Fixing deployment induced by FreeImage packaging
* New version of the segment detector in Scribo module.
* Update range-v3 to version 0.12.0
* Fix horizontal cut issue
* Fix Conan recipes due to update to 1.56.0
* Update according to Conan v2
* Fix conan create on GCC 10
* feat(scribo): update
* Fix CI to publish into public/${branch_name} for all branches.


conanfile.py
from conan import ConanFile
from conan.tools.build import check_min_cppstd
from conan.tools.cmake import CMakeDeps, CMakeToolchain, CMake
from conan.tools.files import rmdir
from conans.errors import ConanInvalidConfiguration

import os

required_conan_version = ">=2.0.1"

class Pylene(ConanFile):
    name = "pylene"
    version = "head"
    license = "MPL v2"
    url = "https://gitlab.lre.epita.fr/olena/pylene"
    description = "C++ Generic Image Processing Library."
    settings = "os", "compiler", "arch", "build_type"
    options = {
        "shared": [True, False],
        "fPIC": [True, False],
    }
    default_options = {
        "shared": False,
        "fPIC": False,
        "boost/*:header_only": True,
        "freeimage/*:shared": True
    }

    exports_sources = ["pylene/*", "cmake/*", "CMakeLists.txt", "LICENSE"]


    requires = [
        "cfitsio/4.1.0@lrde/stable",
        "freeimage/3.18.0@lrde/stable"
    ]

    def build_requirements(self):
        self.test_requires("gtest/1.11.0", options={"shared": False})
        self.test_requires("benchmark/1.7.1")

    def requirements(self):
        self.requires("range-v3/0.12.0", transitive_headers=True)
        self.requires("fmt/9.0.0", transitive_headers=True, transitive_libs=True)
        self.requires("onetbb/2021.7.0", transitive_headers=True)
        self.requires("xsimd/7.4.6", transitive_headers=True)
        self.requires("eigen/3.4.0", transitive_headers=True)
        self.requires("boost/1.81.0", transitive_headers=True)
        self.requires("zlib/1.2.13", override=True)
        self.requires("libwebp/1.2.4", override=True)
        

    def _check_configuration(self):
        check_min_cppstd(self, "20")
        
        if self.settings.os == "Linux":
            accepted_compilers = ["gcc", "clang"]
            accepted_compilers_dict = {
                "gcc": [str(i) for i in range(10, 13)],
                "clang": [str(i) for i in range(11, 16)]
            }
            if not str(self.settings.compiler) in accepted_compilers or not str(self.settings.compiler.version) in accepted_compilers_dict[str(self.settings.compiler)]:
                accepted_e = [f"{comp}-{ver}" for comp in accepted_compilers for ver in accepted_compilers_dict[comp]]
                raise ConanInvalidConfiguration(f"Compiler {self.settings.compiler} version {self.settings.compiler} not handled on Linux (Accepted: {accepted_e})")

        if self.settings.compiler in ["gcc", "clang"] and tools.Version(self.settings.compiler.version) < 10:
            raise ConanInvalidConfiguration("Invalid compiler version for {} (Got {}, should be greater or equal to 10)".format(self.settings.compiler, self.settings.compiler.version))

            if not str(self.settings.compiler) in accepted_compilers or not str(self.settings.compiler.version) in accepted_compilers_dict[str(self.settings.compiler)]:
                accepted_e = [f"{comp}-{ver}" for comp in accepted_compilers for ver in accepted_compilers_dict[comp]]
                raise ConanInvalidConfiguration(f"Compiler {self.settings.compiler} version {self.settings.compiler} not handled on Linux (Accepted: {accepted_e})")

    def _conditional_delete_fPIC(self, cond: bool):
        if cond and self.options.get_safe("fPIC") is not None:
            del self.options.fPIC

    def validate(self):
        self._check_configuration()

    def config_options(self):
        self._conditional_delete_fPIC(self.settings.os == "Windows")

    def configure(self):
        self._conditional_delete_fPIC(self.options.shared)

    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()
        tc = CMakeDeps(self)
        tc.generate()

    def layout(self):
        self.folders.source = "."
        self.folders.build = "build"
        self.folders.generators = "build"
        # Core component
        self.cpp.package.components["core"].libs = ["Pylene-core"]
        self.cpp.package.components["core"].libdirs = ["lib"]
        self.cpp.package.components["core"].includedirs = ["include"]
        self.cpp.source.components["core"].includedirs = ["pylene/include"]
        self.cpp.build.components["core"].libdirs = ["pylene"]

        # Scribo component
        self.cpp.package.components["scribo"].libs = ["Pylene-scribo"]
        self.cpp.package.components["scribo"].libdirs = ["lib"]
        self.cpp.package.components["scribo"].includedirs = ["include"]
        self.cpp.source.components["scribo"].includedirs = ["pylene/include"]
        self.cpp.build.components["scribo"].libdirs = ["pylene"]


        # IO component (FreeImage)
        self.cpp.package.components["io-freeimage"].libs = ["Pylene-io-freeimage"]
        self.cpp.package.components["io-freeimage"].libdirs = ["lib"]
        self.cpp.package.components["io-freeimage"].includedirs = ["include"]
        self.cpp.source.components["io-freeimage"].includedirs = ["pylene/include"]
        self.cpp.build.components["io-freeimage"].libdirs = ["pylene"]


        # IO component (cfitsio)
        self.cpp.package.components["io-fits"].libs = ["Pylene-io-fits"]
        self.cpp.package.components["io-fits"].libdirs = ["lib"]
        self.cpp.package.components["io-fits"].includedirs = ["include"]
        self.cpp.source.components["io-fits"].includedirs = ["pylene/include"]
        self.cpp.build.components["io-fits"].libdirs = ["pylene"]

    def build(self):
        variables = { "PYLENE_BUILD_LIBS_ONLY" : "YES" } # Controls what is built, not the toolchain
        cmake = CMake(self)
        cmake.configure(variables)
        cmake.build()
        cmake.install()

    def package(self):
        rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))

    def package_info(self):
        self.cpp_info.builddirs = ["cmake"]
        self.cpp_info.set_property("cmake_target_name", "pylene::pylene")

        # Core component
        self.cpp_info.components["core"].requires = ["range-v3::range-v3", "fmt::fmt", "onetbb::onetbb", "xsimd::xsimd", "boost::headers"]
        self.cpp_info.components["core"].libs = ["Pylene-core"]
        self.cpp_info.components["core"].includedirs = ["include"]

        # Scribo component
        self.cpp_info.components["scribo"].requires = ["core", "eigen::eigen3"]
        self.cpp_info.components["scribo"].libs = ["Pylene-scribo"]
        self.cpp_info.components["scribo"].includedirs = ["include"]

        # IO component (FreeImage)
        self.cpp_info.components["io-freeimage"].requires = ["core", "freeimage::FreeImage"]
        self.cpp_info.components["io-freeimage"].libs = ["Pylene-io-freeimage"]
        self.cpp_info.components["io-freeimage"].includedirs = ["include"]

        # IO component (cfitsio)
        self.cpp_info.components["io-fits"].requires = ["core", "cfitsio::cfitsio"]
        self.cpp_info.components["io-fits"].libs = ["Pylene-io-fits"]
        self.cpp_info.components["io-fits"].includedirs = ["include"]
